controller/resource.go

```diff
@@ -71,6 +71,89 @@ func (rr *ResourceRepo) Add(r *ct.Resource) error {
 	return tx.Commit()
 }
 
+func (rr *ResourceRepo) AddApp(resourceID, appID string) (*ct.Resource, error) {
+	row := rr.db.QueryRow("resource_select", resourceID)
+	r, err := scanResource(row)
+	if err != nil {
+		return nil, err
+	}
+
+	r.Apps = append(r.Apps, appID)
+
+	tx, err := rr.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+	{
+		var row postgres.Scanner
+		if idPattern.MatchString(appID) {
+			row = tx.QueryRow("app_resource_insert_app_by_name_or_id", appID, appID, r.ID)
+		} else {
+			row = tx.QueryRow("app_resource_insert_app_by_name", appID, r.ID)
+		}
+	scan:
+		if err := row.Scan(&r.Apps[len(r.Apps)-1]); err != nil {
+			if postgres.IsUniquenessError(err, "") {
+				tx, err = rr.db.Begin()
+				if err != nil {
+					return nil, err
+				}
+				if idPattern.MatchString(appID) {
+					row = tx.QueryRow("app_resource_undelete_app_by_name_or_id", appID, appID, r.ID)
+				} else {
+					row = tx.QueryRow("app_resource_undelete_app_by_name", appID, r.ID)
+				}
+				goto scan
+			}
+			tx.Rollback()
+			return nil, err
+		}
+	}
+	if err := createEvent(tx.Exec, &ct.Event{
+		AppID:      appID,
+		ObjectID:   r.ID,
+		ObjectType: ct.EventTypeResource,
```

Comment:
This is following the pattern used elsewhere of sending an event of the same type as at creation each time the item updated.